Makefile

```diff
@@ -1,8 +1,37 @@
-test:
+.PHONY: test lint docgen clean
+
+DEPS_DIR := deps
+PLENARY_DIR := $(DEPS_DIR)/plenary.nvim
+DEVICONS_DIR := $(DEPS_DIR)/nvim-web-devicions
+
+define git_clone_or_pull
+@mkdir -p $(dir $1)
+@if [ ! -d "$1" ]; then \
+	git clone --depth 1 $2 $1; \
+else \
+	git -C "$1" pull; \
+fi
+endef
+
+$(DEPS_DIR):
+	@mkdir -p $@
+
+plenary: | $(DEPS_DIR)
+	$(call git_clone_or_pull,$(PLENARY_DIR),https://github.com/nvim-lua/plenary.nvim)
+
+test-deps: plenary | $(DEPS_DIR)
+	$(call git_clone_or_pull,$(DEVICONS_DIR),https://github.com/nvim-tree/nvim-web-devicons)
+
+test: plenary
```

Comment:
I just recently had a PR from a contributor who couldn't run `make test` since their directory structure wasn't compatible with the existing `minimal_init.lua` so I'm not sure if I agree with the sentiment that this doesn't benefit anyone.
And local tests now depend on whatever version of plenary you currently have (eg. if someone's pinned plenary on some old commit). I think this is less consistent with the testing we do in the CI. 

This gets even worse for docgen. Even if we didn't have the constraint of the pinned version of tree-sitter-lua not working on nightly, this is a dependency contributors have to manage anyways with if they want to docgen locally. Which nobody does currently largely due to this limitation even before docgen broke on nightly.

Reply:
nobody ever ran docgen locally. It was always documented in CONTRIBUTING and people always had questions especially if CI for some reason didnt ran. thats why i started just checking out finished PRs and running it myself.

> I think this is less consistent with the testing we do in the CI.

Doesnt matter, we have CI that says it either works or not. Same for stylua, or luacheck. if you have an old version for that, CI also fails. Should we now introduce nix to work around this issue (for the 0.1% of linux ppl that use nix). I dont think so.

You can never satisfy everyone and having less stuff to maintain and not being relying on a internet connection just to run tests, seems just weird. Each make test run requires this now, which is just unnecessary.